File: wordLadder2.go
package main

func findLadders(beginWord string, endWord string, wordList []string) [][]string {
	wordSet := make(map[string]bool, len(wordList))
	for _, w := range wordList {
		wordSet[w] = true
	}
	if !wordSet[endWord] {
		return [][]string{}
	}

	parents := make(map[string][]string) // child -> parents
	level := map[string]struct{}{beginWord: {}}
	visited := map[string]struct{}{beginWord: {}}
	found := false

	for len(level) > 0 && !found {
		nextLevel := map[string]struct{}{}
		for w := range level {
			for _, v := range neighbors(w, wordSet) {
				if _, seen := visited[v]; seen {
					continue
				}
				parents[v] = append(parents[v], w)
				if _, in := nextLevel[v]; !in {
					nextLevel[v] = struct{}{}
				}
				if v == endWord {
					found = true
				}
			}
		}
		for v := range nextLevel {
			visited[v] = struct{}{}
		}
		level = nextLevel
	}

	if _, ok := parents[endWord]; !ok {
		return [][]string{}
	}

	var res [][]string
	path := []string{endWord}
	var dfs func(string)
	dfs = func(word string) {
		if word == beginWord {
			rev := make([]string, len(path))
			for i := range path {
				rev[i] = path[len(path)-1-i]
			}
			res = append(res, rev)
			return
		}
		for _, p := range parents[word] {
			path = append(path, p)
			dfs(p)
			path = path[:len(path)-1]
		}
	}
	dfs(endWord)
	return res
}

func neighbors(s string, dict map[string]bool) []string {
	res := []string{}
	b := []byte(s)
	for i := 0; i < len(b); i++ {
		orig := b[i]
		for c := byte('a'); c <= 'z'; c++ {
			if c == orig {
				continue
			}
			b[i] = c
			if dict[string(b)] {
				res = append(res, string(b))
			}
		}
		b[i] = orig
	}
	return res
}

File: wordLadder2.py
from collections import defaultdict, deque
from typing import List

class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        word_set = set(wordList)
        if endWord not in word_set:
            return []

        parents = defaultdict(list)  # child -> list of parents on shortest layers
        level = {beginWord}
        visited = set([beginWord])
        found = False

        while level and not found:
            next_level = set()
            for word in level:
                wchars = list(word)
                for i in range(len(wchars)):
                    orig = wchars[i]
                    for c in map(chr, range(ord('a'), ord('z') + 1)):
                        if c == orig: 
                            continue
                        wchars[i] = c
                        nxt = ''.join(wchars)
                        if nxt in word_set and nxt not in visited:
                            parents[nxt].append(word)
                            if nxt not in next_level:
                                next_level.add(nxt)
                            if nxt == endWord:
                                found = True
                    wchars[i] = orig
            visited |= next_level
            level = next_level

        if endWord not in parents:
            return []

        res, path = [], [endWord]

        def dfs(word: str):
            if word == beginWord:
                res.append(path[::-1])
                return
            for p in parents[word]:
                path.append(p)
                dfs(p)
                path.pop()

        dfs(endWord)
        return res
